Move get involved document

sairamya93 opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

To handle the rendering of get_involved document type from government_frontend to frontend as part of [RFC 175](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-175-frontend-fewer-apps.md)

## Why

https://trello.com/c/mXj8oTI3/349-move-document-type-getinvolved-from-government-frontend-to-frontend, [Jira issue PNP-8568](https://gov-uk.atlassian.net/browse/PNP-8568)

## How

To add a scoped route for, a controller, a model and a view. Here we split up the methods in the original presenter file into Presenter and Model logic, with the minimum functionality required to get Get Involved document working.

## Screenshots?

Get Involved page from 
http://frontend.dev.gov.uk/government/get-involved and https://www.gov.uk/government/get_involved

<img width="1468" alt="Screenshot 2024-10-21 at 15 33 26" src="https://github.com/user-attachments/assets/0d869351-8e48-4b86-9b7d-f7aa99772df4">

<img width="1487" alt="Screenshot 2024-10-21 at 15 59 28" src="https://github.com/user-attachments/assets/f66da8d2-796c-4a80-80ee-3de54b941746">

<img width="1414" alt="Screenshot 2024-10-21 at 16 02 16" src="https://github.com/user-attachments/assets/d694a68b-1db1-49c3-8972-b4363be62e20">

<img width="1413" alt="Screenshot 2024-10-21 at 16 03 03" src="https://github.com/user-attachments/assets/13c375a4-5359-421e-a05c-dbe2d8c61137">

<img width="1420" alt="Screenshot 2024-10-21 at 16 08 10" src="https://github.com/user-attachments/assets/ba45c870-5ef0-4012-9a58-407c8347c755">

sairamya93 committed:
Update README.md and update govuk_examples.yml

sairamya93 committed:
Add views page for get-involved

 - Added main wrapper
 - Updated with the modified translation keys

Commit audit trail:
 - https://github.com/alphagov/government-frontend/blob/8c2cdedc7aa3503b343383599f0bdd80eae46dab/app/views/content_items/get_involved.html.erb

KludgeKML commented on app/models/get_involved.rb:
```diff
@@ -0,0 +1,48 @@
+  attr_reader :take_part_pages
+
+  def initialize(content_store_response, override_content_store_hash: nil)
+    super(content_store_response, override_content_store_hash:)
```
For the sake of simplicity, if you call super without any parameters it takes the parameters from the calling method. 

So you could replace `super(content_store_response, override_content_store_hash:)` with just `super`

KludgeKML commented on app/models/content_item.rb:
```diff
@@ -1,6 +1,6 @@
 class ContentItem
```
You could cut out this whole commit, since you're overwriting it later.

leenagupte commented on app/models/get_involved.rb:
```diff
@@ -5,4 +5,44 @@ def initialize(content_store_response, override_content_store_hash: nil)
+    GdsApi.search.search(query)["results"].first
+  end
+
+  def time_until_closure(consultation)
```
This method has moved about a bit. Considering that's looking up specific "get_involved" translation keys it should be in the presenter.

leenagupte commented on .byebug_history:
This file shouldn't have been commited.

KludgeKML replied:
Gone now

leenagupte commented on app/presenters/get_involved_presenter.rb:
```diff
@@ -1,9 +1,5 @@
-  #   @content_item = content_item
-  # end
-
+  
```
It's good that this has been deleted, but this should be fixed up to the previous commit.

leenagupte commented on .gitignore:
```diff
@@ -18,6 +18,7 @@ spec/reports/pacts
 .yarn/cache/
 .yarn/install-state.gz
 .nyc_output
+.byebug_history
```
I think that as this change is not related to Get Involved pages, it should either be in it's own commit, or even better, added in a separate PR.

leenagupte commented on spec/requests/get_involved_spec.rb:
```diff
@@ -34,7 +34,7 @@
   end
 
-  context "when the consultation is closed" do        
+  context "when the consultation is closed" do
```
Lint errors should be fixed up to the commit they first occurred in.

leenagupte commented on spec/requests/get_involved_spec.rb:
```diff
@@ -15,6 +16,29 @@
   end
 end
 
+RSpec.describe "Time until closure" do
```
I don't think we've documented this well, but the convention we have is for the top-level descriptor to use `RSpec.describe` and all of the other blocks to in the test file to either use `describe` if describing a method name, or `context` if describing a scenario. I think for this test only the "when the consultation is closed" context is needed.

leenagupte replied:
Good spot! Thanks

leenagupte reviewed (commented):
I think this is a really good start.

The Get Involved pages turned out to much more complicated then at first realised due to the need to integrate with search. That means that the original presenter in government-frontend has a lot of tightly-coupled model and presentation logic.

You've made a valiant attempt to sort this out, but the to-ing and fro-ing about where a method should go has left the commit history for this PR in a bit of a mess. With these consolidating app PRs the commit history is more important than usual because we need to sign post future developers to where the changes they're looking at originated from.

I think the request and system tests are in a good place, that will help with sorting out the model and presenter.

I think the next steps should be to sort out the nesting in the locale files, and fix up all of the lint errors, and accidentally commited files.

For the model and presenter, I think the best option is to squash those changes into one commit, and then it'll be easier to separate out the methods into their individual parts. I do think those methods can't just be lifted and shifted from the original presenter. which is a real shame.

A last thing to check is the meta tags and make sure that they have been added correctly. The easiest way to do that is to include the shared example in the [system tests](https://github.com/alphagov/frontend/blob/890c15a940fd98ba3b2fd069aae882376eaff064/spec/system/help_page_spec.rb#L2).

I think rebasing and sorting out the model and presenter is going to be quite large job and I'm happy to pair with you on this.

sairamya93 committed:
Add request and system test cases for get_involved

- Add test for meta_tags

sairamya93 committed:
Add model and presenter test cases

sairamya93 commented on app/presenters/get_involved_presenter.rb:
```diff
@@ -0,0 +1,107 @@
+class GetInvolvedPresenter < ContentItemPresenter
```
Updated the presenter and added the link for the original get_involved presenter

sairamya93 committed:
Update locales for Get_Involved

 - Ran rake "consolidation:copy_translation[get_involved,formats.get_involved]" to
update the translation keys

sairamya93 committed:
Add route and controller for Get_Involved

sairamya93 committed:
Add Model and Presenter files to handle the get involved data

- Decoupled the methods to represent the model and presentation logic
- Updated startup.sh for search
- Modify presenter with the updated translation keys

Commit audit trail:
 - https://github.com/alphagov/government-frontend/blob/8c2cdedc7aa3503b343383599f0bdd80eae46dab/app/presenters/get_involved_presenter.rb

KludgeKML commented on spec/requests/get_involved_spec.rb:
```diff
@@ -0,0 +1,38 @@
+RSpec.describe "Get Involved" do
+  before do
+    stub_content_store_has_item("/government/get-involved", GovukSchemas::Example.find("get_involved", example_name: "get_involved"))
+    stub_search_query(query: hash_including(filter_content_store_document_type: "open_consultation"), response: { "results" => [], "total" => 83 })
+    stub_search_query(query: hash_including(filter_content_store_document_type: "closed_consultation"), response: { "results" => [], "total" => 110 })
+    stub_search_query(query: hash_including(filter_content_store_document_type: "consultation_outcome"), response: { "results" => [consultation_result] })
```
I think we want the actual results in the response, which stub_any_search_post doesn't do?

leenagupte replied:
This request test isn't testing the search results, so they're not really needed. I don't think all of that search stubbing is required when just the response and the page title is being tested.

KludgeKML reviewed (approved):
LGTM

leenagupte commented on spec/presenter/get_involved_spec.rb:
```diff
@@ -0,0 +1,111 @@
+RSpec.describe GetInvolved do
+  ## Mock data for testing
```
Minor: This comment isn't needed.

leenagupte commented on spec/models/get_involved_spec.rb:
Minor, but could this tests be moved to the commit following the addition of the model and presenter?

leenagupte commented on app/presenters/get_involved_presenter.rb:
```diff
@@ -0,0 +1,59 @@
+
+  def time_until_next_closure
+    days_left = (content_item.next_closing_consultation["end_date"].to_date - Time.zone.now.to_date).to_i
+    case days_left
+    when :negative?.to_proc
+      I18n.t("formats.get_involved.closed")
+    when :zero?.to_proc
+      I18n.t("formats.get_involved.closing_today")
+    when 1
```
This case statement is pretty hideous. I know you've copied it verbatim from government-frontend, which was the correct thing to do, but perhaps it could be refactored in another commit to use "if/else"?

i.e.

```ruby
if days_left.negative?
  ...
elsif days_left.zero?
  ...
elsif days_left == 1
 ...
else
 ...
end
```

leenagupte commented on config/locales/az.yml:
```diff
@@ -388,6 +388,47 @@ az:
   formats:
     find_my_nearest:
       valid_postcode_no_locations:
+    get_involved:
```
Thanks for moving these ⭐

leenagupte commented on spec/system/get_involved_spec.rb:
```diff
@@ -0,0 +1,60 @@
+RSpec.describe "Get Involved" do
+  it_behaves_like "it has meta tags", "get_involved", "/government/get-involved"
```
This is great!

leenagupte reviewed (commented):
The changes look great! 🎉 

I've added a couple of inline requests for a bit of tidy-up, but it's nothing major. You've done the lions share and it looks good. 

The commit audit trails back to government-frontend looks good too.

leenagupte commented on spec/requests/get_involved_spec.rb:
```diff
@@ -0,0 +1,15 @@
+RSpec.describe "Get Involved" do
+  before do
+    stub_content_store_has_item("/government/get-involved", GovukSchemas::Example.find("get_involved", example_name: "get_involved"))
+    stub_request(:get, /\A#{Plek.new.find('search-api')}\/search.json/).to_return(body: { "results" => [], "total" => 83 }.to_json)
```
This is much better! Thanks for getting rid of the extraneous stubbing in this test 🎉

leenagupte reviewed (approved):
This looks great! 👍 

Thanks for persevering with all of the rebasing.

I've got the page running on Heroku too and it looks good: https://govuk-frontend-app-pr-4305.herokuapp.com/government/get-involved

The commit structure is great, and the audit trail back to government-frontend looks complete. 🎉 

There are two follow-up tasks to do as part of making these changes live:

1. Get the pages looked over by a frontend dev to make sure nothing has been missed. Last time they caught that the `main` tag was missing.
2. Now that frontend relies on `search-api` that dependency should be added to the govuk-docker config, startup script, and app.json (for Heroku). These can be updated in follow-up PRs.

sairamya93 committed:
Add search-api in app.config
